### PR TITLE
Simplify match patterns in coercion rules

### PR DESCRIPTION
**Pull request by findepi**
Remove conditions where unnecessary.
Refactor to improve readability.

### CODE REVIEW

File: datafusion/expr/src/type_coercion/functions.rs [hunk 1]
```diff
@@ -602,89 +602,48 @@ fn coerced_from<'a>(
...
-        (Timestamp(TimeUnit::Nanosecond, None), _)
-            if matches!(
-                type_from,
-                Null | Timestamp(_, None) | Date32 | Utf8 | LargeUtf8
-            ) =>
-        {
-            Some(type_into.clone())
-        }
-        (Interval(_), _) if matches!(type_from, Utf8 | LargeUtf8) => {
+        (Int8, Null | Int8) => Some(type_into.clone()),
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/functions.rs:605]: 😍

**Review by alamb — APPROVED** [review]
Love it -- thank you @findepi

**Review by jayzhan211 — APPROVED** [review]
👍

**Comment by jayzhan211** [comment]
Thanks @findepi @alamb

**Pull request merged**